src/Rings/groebner.jl

```diff
@@ -1221,3 +1243,124 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
 	dim(I) == 0 || error("Dimension of ideal must be zero.")
 	I.gb[destination_ordering] = _fglm(G, destination_ordering)
 end
+
+
+@doc Markdown.doc"""
+    groebner_basis_hilbert_driven(I::MPolyIdeal{P};
+                                  ordering::MonomialOrdering,
+                                  complete_reduction::Bool = false) where {P <: MPolyElem_dec}
+
+
+Compute a Gröbner basis of `I` with respect to `ordering` using a
+Hilbert Series driven method as follows: If a Gröbner basis for `I` is
+present, compute the Hilbert series of `I` and use it to optimize the
+Gröbner basis computation for `I` w.r.t. `ordering`. If no Gröbner
+basis for `I` is present compute the Hilbert series for `I` if the
+base field of `I` has positive characteristic, otherwise compute the
+Hilbert series for `I` modulo a randomly chosen prime. Use the
+resulting Hilbert series to optimize the Gröbner basis computation for
+`I` w.r.t. `ordering`.
+
+`I` must be given by generators homogeneous w.r.t. `weights`.
+
+# Examples
+```jldoctest
+julia> R, (x, y, z) = grade(PolynomialRing(QQ, ["x", "y", "z"])[1]);
+
+julia> I = ideal(R, [x^2 + y*z, x*y - y*z]);
+
+julia> groebner_basis_hilbert_driven(I, ordering = deglex(R), complete_reduction = true)
+Gröbner basis with elements
+1 -> x*y - y*z
+2 -> x^2 + y*z
+3 -> y^2*z + y*z^2
+with respect to the ordering
+deglex([x, y, z])
+```
+"""
+
+function groebner_basis_hilbert_driven(I::MPolyIdeal{P};
+                                       ordering::MonomialOrdering,
+                                       complete_reduction::Bool = false) where {P <: MPolyElem_dec}
+  
+  all(p -> is_homogeneous(p), gens(I)) || error("I must be given by generators homogeneous with respect to its underlying ring.")
+	isa(coefficient_ring(base_ring(I)), AbstractAlgebra.Field) || error("The underlying coefficient ring of I must be a field.")
+  is_global(ordering) || error("Destination ordering must be global.")
+	haskey(I.gb, ordering) && return I.gb[ordering]
+  if isempty(I.gb) && iszero(characteristic(base_ring(I)))  
+    while true
+      r = rand(2^15:2^16)
+      p = Hecke.next_prime(r)
+        
+      base_field = GF(p)
+      ModP, _ = grade(PolynomialRing(base_field, "x" => 1:ngens(base_ring(I)))[1],
+                      _extract_weights(base_ring(I)))
+      I_mod_p_gens = Vector{elem_type(ModP)}(undef, length(gens(I))) 
+      try
+        # we have to coerce to ModP like this because the base ring of `I` is
+        # graded
+        for (i, f) in enumerate(gens(I))
+          ctx = MPolyBuildCtx(ModP)
+          for (c, m) in zip(coefficients(f), exponents(f))
+            push_term!(ctx, base_field(c), m)
+          end
+          I_mod_p_gens[i] = finish(ctx)
+        end
```

Comment:
I have not tested this, but: can this perhaps be simplified to
```suggestion
        for (i, f) in enumerate(gens(I))
          I_mod_p_gens[i] = map_coefficients(base_field, f; parent=ModP)
        end
```
or even
```suggestion
        I_mod_p_gens = [map_coefficients(base_field, f; parent=ModP) for f in gens(I)]
```

Reply:
It can! Added in https://github.com/oscar-system/Oscar.jl/pull/1821/commits/ea977e7db3ee648e386f4a30508a24ae9ecab226